### PR TITLE
Bug/etcm 797

### PR DESCRIPTION
**Pull request by bsuieric**
# Description

https://jira.iohk.io/browse/ETCM-797

# Proposed Solution

calculate start epoch block, and calculate seed based on the epoch length and start epoch block

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/jsonrpc/DebugService.scala [hunk 1]
```diff
@@ -41,4 +44,27 @@ class DebugService(peerManager: ActorRef, etcPeerManager: ActorRef) {
...
+              GetBlockHeaders(Left(req.number - 1), 2, 0, false),
+              peerId
+            ),
+            peerManager
+          )
+        }
+      }
+    } yield Right(GetSpecificBlockResponse())
+
+//    Task(Right(GetSpecificBlockResponse()))
```
Comment (bsuieric) [inline src/main/scala/io/iohk/ethereum/jsonrpc/DebugService.scala]: remove comment

File: src/main/scala/io/iohk/ethereum/jsonrpc/DebugService.scala [hunk 2]
```diff
@@ -41,4 +44,27 @@ class DebugService(peerManager: ActorRef, etcPeerManager: ActorRef) {
       .askFor[PeerInfoResponse](EtcPeerManagerActor.PeerInfoRequest(peer))
       .map(resp => resp.peerInfo)
   }
+
+  def getSpecificBlock(req: GetSpecificBlockRequest): ServiceResponse[GetSpecificBlockResponse] = {
```
Comment (bsuieric) [inline src/main/scala/io/iohk/ethereum/jsonrpc/DebugService.scala]: 👌

**Review by jvdp — APPROVED** [review]

File: src/main/scala/io/iohk/ethereum/consensus/pow/EthashUtils.scala [hunk 3]
```diff
@@ -63,12 +63,18 @@ object EthashUtils {
...
 
-  def seed(blockNumber: Long): ByteString = {
-    val epoch = epochBeforeEcip1099(blockNumber)
+  def seed(blockNumber: Long, ecip1099ActivationBlock: Long): ByteString = {
+    val epoch: Long = epochBeforeEcip1099(startBlock(blockNumber, ecip1099ActivationBlock))
     (BigInt(0) until epoch)
       .foldLeft(ByteString(Hex.decode("00" * 32))) { case (b, _) => kec256(b) }
   }
 
+  def startBlock(blockNumber: Long, ecip1099ActivationBlock: Long): Long =
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/consensus/pow/EthashUtils.scala]: this is quite confusing - why such a convoluted computation?
can `calcEpochLength(blockNumber, ecip1099ActivationBlock)` change 
in this multiplication `calcEpochLength(blockNumber, ecip1099ActivationBlock)) * calcEpochLength(blockNumber, ecip1099ActivationBlock)`?

perhaps
```scala
val epochLength = calcEpochLength(blockNumber, ecip1099ActivationBlock)
((blockNumber / epochLength) * epochLength) + 1
```
or maybe even more explicitly:
```scala
val epochLength = calcEpochLength(blockNumber, ecip1099ActivationBlock)
val epoch: Long = blockNumber / epochLength
(epoch * epochLength + 1) / EPOCH_LENGTH_BEFORE_ECIP_1099
```
to avoid this indirection:
`epochBeforeEcip1099(startBlock(blockNumber, ecip1099ActivationBlock))`
?

Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/consensus/pow/EthashUtils.scala]: Deduplicating that `calcEpochLength(...)` call sounds good, but I'm not sure your other proposals are equivalent.

File: src/main/scala/io/iohk/ethereum/consensus/pow/EthashDAGManager.scala [hunk 4]
```diff
@@ -13,11 +13,11 @@ class EthashDAGManager(blockCreator: EthashBlockCreator) extends Logger {
   var currentEpochDagSize: Option[Long] = None
   var currentEpochDag: Option[Array[Array[Int]]] = None
 
-  def calculateDagSize(blockNumber: Long, epoch: Long): (Array[Array[Int]], Long) = {
+  def calculateDagSize(blockNumber: Long, epoch: Long, ecip1099ActivationBlock: Long): (Array[Array[Int]], Long) = {
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/consensus/pow/EthashDAGManager.scala]: wouldn't it make sense to make this parameter a characteristic of the manager itself? this cannot change during the run, and passing it all around seems tedious.

File: src/main/scala/io/iohk/ethereum/consensus/pow/EthashUtils.scala (hunk 3 above)
Comment (bsuieric) [reply src/main/scala/io/iohk/ethereum/consensus/pow/EthashUtils.scala]: I agree, maybe also the name should change to startEpochBlock, or add a comment, so we have more context when we look at it

File: src/main/scala/io/iohk/ethereum/jsonrpc/DebugJsonMethodsImplicits.scala [hunk 5]
```diff
@@ -12,4 +19,19 @@ object DebugJsonMethodsImplicits extends JsonMethodsImplicits {
       def encodeJson(t: ListPeersInfoResponse): JValue =
         JArray(t.peers.map(a => JString(a.toString)))
     }
+
+  implicit val debug_getSpecificBlockRequest: JsonMethodCodec[GetSpecificBlockRequest, GetSpecificBlockResponse] =
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/jsonrpc/DebugJsonMethodsImplicits.scala]: I think that this endpoint shouldn't be part of this PR.
1) it's not documented or tested
2) it's not part of the bugfix
3) will make it harder to backport to a release branch if the need arises

I'm fine with having this come from a ticket that's fleshed out and has its parameters properly identified, perhaps during a refinement session.

File: src/test/scala/io/iohk/ethereum/consensus/pow/EthashUtilsSpec.scala [hunk 6]
```diff
@@ -18,8 +20,21 @@ class EthashUtilsSpec extends AnyFlatSpec with Matchers with ScalaCheckPropertyC
   val ecip1099forkBlockNumber: Long = 11460000
 
   "Ethash" should "generate correct hash" in {
-    forAll(Gen.choose[Long](0, 15000000L)) { blockNumber =>
-      seed(blockNumber, ecip1099forkBlockNumber) shouldBe seedForBlockReference(blockNumber)
+    val seedEpoch0 = ByteStringUtils.string2hash("0000000000000000000000000000000000000000000000000000000000000000")
+    val seedEpoch1 = ByteStringUtils.string2hash("290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563")
+    val seedEpoch382 = ByteStringUtils.string2hash("d3d0aa11197dcdcfcb3ad3c73d415af47299bddb47fda6081d31d9dd06462f6a")
+    val seedEpoch383 = ByteStringUtils.string2hash("bf532874eb434842e7a3e4acd113fe454541651872760d9b95d11d7f90ca25dc")
```
Comment (dzajkowski) [inline src/test/scala/io/iohk/ethereum/consensus/pow/EthashUtilsSpec.scala:26]: where did these hashes come from? maybe this would be a good inspiration https://github.com/iquidus/ecip-1099-data ?

**Review by dzajkowski — CHANGES_REQUESTED** [review]
Please remove the "debug_" endpoint. We can groom it and bring it in properly.

File: src/test/scala/io/iohk/ethereum/consensus/pow/EthashUtilsSpec.scala (hunk 6 above)
Comment (jvdp) [reply src/test/scala/io/iohk/ethereum/consensus/pow/EthashUtilsSpec.scala:26]: These are the seeds that correspond to the blocks, calculated with the Mantis code found to be working properly (valid blocks are valid.)

But the strings themselves are not so interesting for the purposes of this bug, just that they are the same across different blocks within a (ECIP-1099) epoch. (For the implementation itself the hashes are important of course.)

Comment (jvdp) [reply src/test/scala/io/iohk/ethereum/consensus/pow/EthashUtilsSpec.scala:26]: (Also, the hash being all zeroes for epoch 0 is a good sign?)

File: src/main/scala/io/iohk/ethereum/jsonrpc/DebugJsonMethodsImplicits.scala (hunk 5 above)
Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/jsonrpc/DebugJsonMethodsImplicits.scala]: Fair enough. I think this endpoint might be useful for state sync improvements, and we'll be able to grab it from this PR when that's the case.

File: src/main/scala/io/iohk/ethereum/consensus/pow/EthashUtils.scala (hunk 3 above)
Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/consensus/pow/EthashUtils.scala]: I merged it all into the `seed` function, since it's only used there anyway.

**Review by dzajkowski — APPROVED** [review]
